src/moonlink/src/storage/mooncake_table/snapshot_read_output.rs

```diff
@@ -71,50 +84,104 @@ impl ReadOutput {
             .unwrap();
     }
 
+    /// File ordering should be preserved because the index determines the position
+    /// in the final [`resolved_data_files`] array, which delete or other operations rely on.
+    async fn resolve_remote_files(
+        &mut self,
+        object_storage_cache: Arc<dyn CacheTrait>,
+        filesystem_accessor: Arc<dyn BaseFileSystemAccess>,
+        remote_files_entries: Vec<RemoteFileEntry>,
+        resolved_data_files: &mut [String],
+        cache_handles: &mut Vec<NonEvictableHandle>,
+    ) -> Result<()> {
+        let mut results = stream::iter(remote_files_entries.into_iter())
+            .map(|remote_file_entry| {
+                let cache = object_storage_cache.clone();
+                let fs_accessor = filesystem_accessor.clone();
+                async move {
+                    let result = cache
+                        .get_cache_entry(
+                            remote_file_entry.file_id,
+                            &remote_file_entry.remote_filepath,
+                            fs_accessor.as_ref(),
+                        )
+                        .await;
+                    (
+                        remote_file_entry.index,
+                        remote_file_entry.remote_filepath,
+                        result,
+                    )
+                }
+            })
+            .buffer_unordered(MAX_PARALLEL_OPERATIONS);
+
+        while let Some((index, remote_filepath, result)) = results.next().await {
+            match result {
+                Ok((cache_handle, files_to_delete)) => {
+                    if let Some(cache_handle) = cache_handle {
+                        resolved_data_files[index] = cache_handle.get_cache_filepath().to_string();
+                        cache_handles.push(cache_handle);
+                    } else {
+                        resolved_data_files[index] = remote_filepath;
+                    }
+
+                    self.notify_evicted_files(files_to_delete.into_vec()).await;
+                }
+                Err(e) => {
+                    self.handle_resolution_error(std::mem::take(cache_handles))
```

Comment:
It looks like an issue, if we return early, chances are that un-synced tasks have pinned cache handles, which should be unpinned as well?